libres/lib/enkf/enkf_obs.cpp

```diff
@@ -1154,11 +1096,91 @@ conf_class_type *enkf_obs_get_obs_conf_class(void) {
     return enkf_conf_class;
 }
 
+hash_iter_type *enkf_obs_alloc_iter(const enkf_obs_type *enkf_obs) {
+    return hash_iter_alloc(enkf_obs->obs_hash);
+}
+
 /*
-   Allocates a stringlist of obs target keys which correspond to
-   summary observations, these are then added to the state vector in
-   enkf_main.
+   This function allocates a hash table which looks like this:
+
+     {"OBS_KEY1": "STATE_KEY1", "OBS_KEY2": "STATE_KEY2", "OBS_KEY3": "STATE_KEY3", ....}
+
+   where "OBS_KEY" represents the keys in the enkf_obs hash, and the
+   value they are pointing at are the enkf_state keywords they are
+   measuring. For instance if we have an observation with key "RFT_1A"
+   the entry in the table will be:  ... "RFT_1A":  "PRESSURE", ..
+   since an RFT observation observes the pressure.
+
+   Let us consider the watercut in a well. Then the state_kw will
+   typically be WWCT:P1 for a well named 'P1'. Let us assume that this
+   well is observed both as a normal HISTORY observation from
+   SCHEDULE, and from two separator tests, called S1 and S2. Then the
+   hash table will look like this:
+
+       "WWCT:P1": "WWCT:P1",
+       "S1"     : "WWCT:P1",
+       "S2"     : "WWCT:P1"
+
+
+   I.e. there are three different observations keys, all observing the
+   same state_kw.
 */
+
+hash_type *enkf_obs_alloc_data_map(enkf_obs_type *enkf_obs) {
+    hash_type *map = hash_alloc();
+    hash_iter_type *iter = hash_iter_alloc(enkf_obs->obs_hash);
+    const char *key = hash_iter_get_next_key(iter);
+    while (key != NULL) {
+        obs_vector_type *obs_vector =
+            (obs_vector_type *)hash_get(enkf_obs->obs_hash, key);
+        hash_insert_ref(map, key, obs_vector_get_state_kw(obs_vector));
+        key = hash_iter_get_next_key(iter);
+    }
+    hash_iter_free(iter);
+    return map;
+}
+
+bool enkf_obs_is_valid(const enkf_obs_type *obs) { return obs->valid; }
+
+void enkf_obs_free(enkf_obs_type *enkf_obs) {
+    hash_free(enkf_obs->obs_hash);
+    vector_free(enkf_obs->obs_vector);
+    time_map_free(enkf_obs->obs_time);
+    free(enkf_obs);
+}
+
+time_t enkf_obs_iget_obs_time(const enkf_obs_type *enkf_obs, int report_step) {
+    return time_map_iget(enkf_obs->obs_time, report_step);
+}
+
+bool enkf_obs_has_key(const enkf_obs_type *obs, const char *key) {
+    return hash_has_key(obs->obs_hash, key);
+}
```

Comment:
Could you reorder this code a bit, or put the moving in a different commit? As far as I can tell the code is only moving, it is not changing? Right now the diff is a bit bigger than it needs to be.

Reply:
OK - I removed unnecessary changes to `enkf_obs`